### ISSUE TITLE
Autodetect pruning opportunities 

### ISSUE THREAD

**Issue by fulmicoton** [quickwit-inc/quickwit#823]
Some queries will look like  `some_tag_field:toto AND text:help`

In that case, if `some_tag_field` happens to be a tag, it would be highly beneficial to automatically use it to prune the splits targetted in the query.

**Comment by fulmicoton** [comment]
Everything sounds great

**Comment by fulmicoton** [comment]
As raised by @fmassot and discussed with @evanxg852000 yesterday:

#437 will require use to modify the metastore in order to make it handle more complicated boolean logic than a simple conjunction of tags for tag-pruning.

The current plan is to pass it a list of list of tags, and interpret it as a CNF. 
[[a, b], [c]] becomes [A OR B] AND [C].

In parallel, as pointed out by Francois in the PR, we need a clear plan on what kind of tag pruning we should do when faced with a query like 
`tag:A OR tag:B`

In that case the goal is to extract a boolean Predicate containing only tags as literals, such that matching the query implies P.

For instance,
```
tag:A AND body:text -> tag:A
tag:A OR body:text -> true
(tag:A OR tag:B) AND body:text -> tag:A OR tag:B
```

Our query language contains a NOT operator too... 

It is possible to integrate it in the logic too. Should we ? @fmassot @evanxg852000 @r-zenine 
I suggest we don't. In that case, the extraction logic should look like follows:

A safe way to handle not (although suboptimal) could be to 
treat (NOT E) as if it was "true".

```
NOT(tag:A OR tag:B OR NOT tag:C) AND tag:D -> true AND tag:D -> tag:C
```

It is not optimal because actually
```
NOT(tag:A OR tag:B OR NOT tag:C) AND tag:D => tag:C AND tag:D
```
but we probably don't care.

Am I overthinking this @fmassot ?


TLDR:
- We probably want to land #437 before finishing this.
- The logic that extracts a list of tags from the user input AST is currently broken
- We don't want to just a extract a list of tags, we want to extract predicate over tags. Ideally in the form of a CNF.

**Comment by fmassot** [comment]
Agreed, let's handle the `NOT` operator later
